Edits to tibble pre-announcement blog post (#2)
krlmlr/tidyverse.org: content/articles/2018-11-tibble-2.0.0-pre-announce.Rmd

---
title: "Coming soon: tibble 2.0.0"
slug: tibble-2.0.0-pre-announce
description: >
    The upcoming tibble 2.0.0 release has internal changes relevant to package developers who depend on tibble.
date: 2018-11-21
author: Kirill Müller, Jenny Bryan
photo:
  url: https://unsplash.com/photos/yYawh30qf28
  author: Gabriel Porras
categories: [package]
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(collapse = TRUE, comment = "#>", error = TRUE)
```

> Version 2.0.0 of the [tibble package](https://tibble.tidyverse.org) is almost ready for release. Tibbles are a modern reimagining of the data frame, keeping what time has shown to be effective, and throwing out what is not, with nicer default output too! Grab the development version with:

Author: Have you already bumped the version number? I'm thinking about people implementing the actual minimum version conditions shown in the post.

Reviewer: I guess I should bump before notifying maintainers and publishing the blog post.

```{r eval = FALSE}
devtools::install_github("tidyverse/tibble")
```

We're making a pre-release announcement, because some changes require the attention of maintainers of packages that import or otherwise depend on tibble.

> This post describes how to adapt to the next version of tibble and is also an invitation for maintainers to provide feedback before v2.0.0 is finalized and submitted to CRAN (TODO: WHERE? HOW?).

Author: We need to specify how we want this feedback. Possibilities:

  * We create a tibble issue and send people there.
  * We encourage people to open issues on tibble.
  * A thread on community? Probably not the right venue.

> This post describes how to adapt to the next version of tibble and is also an invitation for maintainers to provide feedback before v2.0.0 is finalized and submitted to CRAN (TODO: WHERE? HOW?).

Reviewer: ```suggestion
This post describes how to adapt to the next version of tibble and is also an invitation for maintainers to provide feedback before v2.0.0 is finalized and submitted to CRAN.
The easiest way to get in touch is to file an issue at https://github.com/tidyverse/tibble/issues (or to comment to an existing one).
```

This blog post is aimed at package developers and those who maintain "production" scripts or apps. A high-level overview of new user-facing features will come in a separate blog post.

## Reverse dependency checks

We ran `R CMD check` for over 3000 CRAN and Bioconductor packages that depend directly or indirectly on the tibble package and compared results obtained with the CRAN versus development version of tibble.

> We notified the maintainers of TODO: FILL IN A NUMBER? affected packages and aim for a CRAN release before Christmas, so the dust has settled in time for [rstudio::conf](https://www.rstudio.com/conference/).

Author: It would be nice to quantify how many packages showed up on the radar.

> We notified the maintainers of TODO: FILL IN A NUMBER? affected packages and aim for a CRAN release before Christmas, so the dust has settled in time for [rstudio::conf](https://www.rstudio.com/conference/).

Reviewer: ```suggestion
We will notify the maintainers of all affected packages (regardless of the check results of their package) and aim for a CRAN release before Christmas, so the dust has settled in time for [rstudio::conf](https://www.rstudio.com/conference/).
```

End users should use the `tibble()` function to construct tibbles.
It checks the input for consistency and makes sure that the returned tibble is valid.

Package developers, however, can also consider the low-level `new_tibble()` constructor. Use `new_tibble()` to quickly construct a tibble from a list if you are very sure that the input is well-formed (i.e., a list of vectors of equal length).
This function also supports the construction of subclasses of tibble through the `class` argument.

In the development version of tibble, the `new_tibble()` constructor is a more faithful implementation of the [design advice for S3 classes given in Advanced R](https://adv-r.hadley.nz/s3.html#s3-classes).
Specifically:

- `new_tibble()` is very fast again and does very little checking itself.
- The new `validate_tibble()` function is responsible for validating the structure of a tibble.

This means that the `nrow` argument to `new_tibble()` is now mandatory.
We are aware that this might be the single most disruptive change, but we think that any guesswork here would be detrimental to stability (especially in corner cases) and that this particular problem is very easy to fix.

> The `nrow` argument already existed in tibble 1.4.2, so code that uses it requires no change and should continue to work.

Reviewer: ```suggestion
The `nrow` argument already existed in tibble 1.4.2, so code that uses it requires no change and should continue to work.
This means you can update your package independently of the tibble 2.0.0 release and do not need a versioned dependency on tibble.
```

Author: The last sentence sort of implies that people need to update but if someone already uses `nrow`, they don't need to update, right? Or, at least, not for `nrow`-releated reasons.

Reviewer: Then maybe:

```suggestion
The `nrow` argument already existed in tibble 1.4.2, so code that uses it requires no change and should continue to work.
If you need to add `nrow` arguments to `new_tibble()` calls, you can do so independently of the tibble v2.0.0 release.
```

Please be aware that `nrow` must be passed as a named argument, because it comes after the ellipsis `...` in the signature. Here are common patterns for setting the `nrow` argument:

```{r}
library(tibble)

x <- data.frame(a = 1)

# Code that lacks `nrow` fails
new_tibble(x)

# Fix by specifying `nrow`
new_tibble(x, nrow = nrow(x)) # if x is a data frame

nrow_x <- NROW(x[[1]]) # if x has at least one column
# nrow_x <- ... # if the number of rows is given elsewhere
new_tibble(x, nrow = nrow_x)
```

> ## Coercion and name repair

Author: I tried to show code and concrete examples as early and as often as possible. This is a lot of your content, but rearranged.

# problematic approach:
# 1. make tibble
# 2. apply nice names
x <- as_tibble(m)
colnames(x) <- letters[1:4]

# better approach that works with tibble v1.4.2 AND dev tibble:
# 1. apply nice names
# 2. make tibble
colnames(m) <- letters[1:4]
as_tibble(m)

```

If possible, we recommend applying your "good" column names prior to calling `as_tibble()`. This creates code that works with tibble 1.4.2 and the development version, which is very appealing. For good examples, see these pull requests to [drake](https://github.com/ropensci/drake/pull/586), [prophet](https://github.com/facebook/prophet/pull/739), and [broom](https://github.com/tidymodels/broom/pull/534).

It is also possible to use the new `.name_repair` argument in `tibble()` and `as_tibble()` (more below) to explicitly declare your intention around column names. This code would require `packageVersion("tibble") >= "2.0.0"`:

```{r}
# Alternative: use new `.name_repair` argument to permit dysfunctional names
m <- cov(unname(iris[-5]))
as_tibble(m, .name_repair = "minimal")

# Alternative: use new `.name_repair` argument to fix dysfunctional names
m <- cov(unname(iris[-5]))
as_tibble(m, .name_repair = "unique")
```

> What is the motivation for this increased attention to column names? The tibble package is offering stronger encouragement for names where each column can be identified by name and, preferably, without having to resort to backticks. Column names that don't meet these requirements are still allowed, but the user needs to permit them explicitly.

Author: I should update my PR to the principles repo and probably merge it soon. We could potentially link to that.

https://github.com/tidyverse/principles/pulls

Reviewer: Let's merge principles before.

See `` ?`name-repair` `` for more details.

Packages that are in the business of making tibbles may even want to expose the `.name_repair` argument and pass it through to tibble. For example, this is the approach planned for [readxl](https://readxl.tidyverse.org), which reads rectangular data out of Excel workbooks.

## Deprecation of `validate` in `as_tibble()`

In tibble 1.4.2, `as_tibble()` has a `validate` argument, but its default behaviour value was inconsistent across different methods and there was no equivalent argument for `tibble()`. The `validate` argument is now soft-deprecated and its use will trigger a message (we intend to make this "once per session"). The `validate` argument will eventually be removed.

Here's what `validate` does in the development version of tibble:

```{r}
m <- cov(unname(iris[-5]))
df <- new_tibble(list(a = 5, a = 6), nrow = 1)

# errors, as it should, because columns are unnamed ... but also messages
as_tibble(m, validate = TRUE)

# errors, as it should, because names are duplicated ... but also messages
as_tibble(df, validate = TRUE)

# no error ... but still messages
as_tibble(df, validate = FALSE)
```

> The default behaviour (`.name_repair = "check_unique`) implements `validate = TRUE`. So you can simply remove `validate = TRUE`:

Reviewer: UK/NZ or US spelling?

It's difficult, because for matrices we had auto-naming of the form `V#`, and for tibbles we never checked (default was `validate = FALSE`). Can we rephrase this around suggesting to specify the desired name repair if the code now fails? Maybe it's worthwhile to treat data frames, tibbles and matrices separately?

Reviewer: This still needs attention, I believe, but I can fix later in my branch.

```{r}
as_tibble(m)
as_tibble(df)
```

Here are alternatives to `validate = FALSE`, if you know that `packageVersion("tibble") >= "2.0.0"`:

```{r}
# permit non-unique names
as_tibble(df, .name_repair = "minimal")

# authorize creation of unique names
as_tibble(m, .name_repair = "unique")
```

## Deprecation of `tidy_names()` and `set_tidy_names()`

The existing `tidy_names()` and `set_tidy_names()` functions are soft-deprecated, but remain available, unchanged. In the future, they could go away or take on a new meaning, i.e. implement a different algorithm for name repair. New code should use `.name_repair` instead.

```{r}
df <- new_tibble(list(a = 5, a = 6), nrow = 1)

# these functions are soft-deprecated
tidy_names(names(df))
set_tidy_names(df)

# achieve same via `.name_repair`
as_tibble(df, .name_repair = "universal")
```

tibble's name repair strategies are currently only exposed in `tibble()` and `as_tibble()`, not (yet?) as utility functions that operate on a vector of names.

> 

Reviewer: Should we keep the pointers? Maybe even add links to the pkgdown?

```suggestion
To work around, you can create a zero-row tibble, using e.g. `new_tibble()` and `rlang::rep_named()`.

```

Author: It is pretty unattractive and I think we don't want to encourage people to create their own utilities. I think we could either wait and see if anyone asks or, if you want, include it as an example in the `name-repair` help.
